### PR TITLE
Compositions from JuLie

### PR DESCRIPTION
**Pull request by joschmitt**
Move the compositions from JuLie to OSCAR.
I actually made the iterator call the existing `weak_compositions` iterator; the algorithm in JuLie came down to the same and this way it is less code.

I also moved the `ascending_partitions` out of experimental, named them `ascending_compositions` as discussed at some point and made the function return an iterator. I only kept one of the two implemented algorithms; the runtimes were basically the same.

Related to #3147.

### CODE REVIEW

**Comment by thofma** [comment]
@mjrodgers can you have a (superficial) look?

File: src/Combinatorics/EnumerativeCombinatorics/compositions.jl [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+
+By a composition of `n` we mean a sequence of positive integers whose sum is `n`.
+
+# Examples
+```jldoctest
+julia> C = compositions(4)
+Iterator over the compositions of 4
+
+julia> collect(C)
+8-element Vector{Oscar.Composition{Int64}}:
```
Comment (lgoettgens) [inline src/Combinatorics/EnumerativeCombinatorics/compositions.jl]: The type `Partition` is exported and thus prints as `Vector{Partition{Int64}}` in this context. Maybe export `Combination` and `WeakCombination` as well?

**Comment by mjrodgers** [comment]
> @mjrodgers can you have a (superficial) look?

I have looked it over, everything here looks good to me.

File: src/Combinatorics/EnumerativeCombinatorics/compositions.jl (hunk 1 above)
Comment (joschmitt) [reply src/Combinatorics/EnumerativeCombinatorics/compositions.jl]: Yes, why not.

**Review by lgoettgens — APPROVED** [review]

**Comment by joschmitt** [comment]
In the 5th attempt the last CI job finally avoided 3676. Please merge.

**Pull request merged**